File: api/src/main/java/org/pmiops/workbench/cohortreview/CohortReviewService.java
package org.pmiops.workbench.cohortreview;

import java.sql.Timestamp;
import java.util.List;
import org.pmiops.workbench.cohortreview.util.PageRequest;
import org.pmiops.workbench.db.model.DbCohort;
import org.pmiops.workbench.db.model.DbCohortReview;
import org.pmiops.workbench.db.model.DbParticipantCohortAnnotation;
import org.pmiops.workbench.db.model.DbParticipantCohortStatus;
import org.pmiops.workbench.db.model.DbUser;
import org.pmiops.workbench.model.CohortChartData;
import org.pmiops.workbench.model.CohortReview;
import org.pmiops.workbench.model.CohortStatus;
import org.pmiops.workbench.model.Domain;
import org.pmiops.workbench.model.ModifyParticipantCohortAnnotationRequest;
import org.pmiops.workbench.model.ParticipantChartData;
import org.pmiops.workbench.model.ParticipantCohortAnnotation;
import org.pmiops.workbench.model.ParticipantCohortStatus;
import org.pmiops.workbench.model.ParticipantData;
import org.pmiops.workbench.model.Vocabulary;

public interface CohortReviewService {

  /** Find the {@link DbCohort} for the specified cohortId. */
  DbCohort findCohort(long workspaceId, long cohortId);

  /** Find the {@link DbCohortReview} for the specified cohortId and cdrVersionId. */
  CohortReview findCohortReview(Long cohortId, Long cdrVersionId);

  /** Find the {@link DbCohortReview} for the specified cohortReviewId. */
  CohortReview findCohortReview(Long cohortReviewId);

  /** Find the {@link CohortReview} for the specified workspaceId and cohortReviewId. */
  CohortReview findCohortReviewForWorkspace(Long workspaceId, Long cohortReviewId);

  /** Delete the specified cohort review. */
  void deleteCohortReview(Long cohortReviewId);

  /** Find the {@link DbCohortReview} for the specified ns and firecloudName. */
  List<CohortReview> getRequiredWithCohortReviews(String ns, String firecloudName);

  /** Save the specified {@link CohortReview}. */
  CohortReview saveCohortReview(CohortReview cohortReview, DbUser creator);

  /**
   * Save the {@link DbCohortReview} as well as the collection of {@link DbParticipantCohortStatus}.
   */
  void saveFullCohortReview(
      CohortReview cohortReview, List<DbParticipantCohortStatus> participantCohortStatuses);

  CohortReview updateCohortReview(
      CohortReview cohortReview, Long cohortReviewId, Timestamp lastModified);

  /** Update the specified {@link ParticipantCohortStatus}. */
  ParticipantCohortStatus updateParticipantCohortStatus(
      Long cohortReviewId, Long participantId, CohortStatus status, Timestamp lastModified);

  /**
   * Find the {@link ParticipantCohortStatus} for the specified cohortReviewId and participantId.
   */
  ParticipantCohortStatus findParticipantCohortStatus(Long cohortReviewId, Long participantId);

  /**
   * Find a list of {@link ParticipantCohortStatus} for the specified cohortReviewId, filtering and
   * paging.
   */
  List<ParticipantCohortStatus> findAll(Long cohortReviewId, PageRequest pageRequest);

  /** Find count for the specified cohortReviewId, filtering and paging. */
  Long findCount(Long cohortReviewId, PageRequest pageRequest);

  /** Save the specified {@link ParticipantCohortAnnotation}. */
  ParticipantCohortAnnotation saveParticipantCohortAnnotation(
      Long cohortReviewId, ParticipantCohortAnnotation participantCohortAnnotation);

  /**
   * Save the {@link ParticipantCohortAnnotation} for the specified participantId, annotationId and
   * {@link ModifyParticipantCohortAnnotationRequest}.
   */
  ParticipantCohortAnnotation updateParticipantCohortAnnotation(
      Long annotationId,
      Long cohortReviewId,
      Long participantId,
      ModifyParticipantCohortAnnotationRequest modifyParticipantCohortAnnotationRequest);

  /**
   * Delete the {@link DbParticipantCohortAnnotation} for the specified annotationId, cohortReviewId
   * and participantId.
   */
  void deleteParticipantCohortAnnotation(
      Long annotationId, Long cohortReviewId, Long participantId);

  /**
   * Find a list of {@link ParticipantCohortAnnotation} for the specified cohortReviewId and
   * participantId.
   */
  List<ParticipantCohortAnnotation> findParticipantCohortAnnotations(
      Long cohortReviewId, Long participantId);

  /** Initialize a {@link CohortReview} */
  CohortReview initializeCohortReview(Long cdrVersionId, DbCohort dbCohort);

  /**
   * Create a list of {@link DbParticipantCohortStatus} for the specified cohort, requestSize and
   * review id.
   */
  List<DbParticipantCohortStatus> createDbParticipantCohortStatusesList(
      DbCohort dbCohort, Integer requestSize, Long cohortReviewId);

  /** Find a list of {@link CohortChartData} for the specified cohort and domain. */
  List<CohortChartData> findCohortChartData(DbCohort dbCohort, Domain domain, int limit);

  /** Find a list of {@link ParticipantChartData} for the specified participant id and domain. */
  List<ParticipantChartData> findParticipantChartData(Long participantId, Domain domain, int limit);

  /** Find participant count. */
  Long findParticipantCount(Long participantId, Domain domain, PageRequest pageRequest);

  /** Find list of {@link ParticipantData} for specifice participant id and domain. */
  List<ParticipantData> findParticipantData(
      Long participantId, Domain domain, PageRequest pageRequest);

  /** Find list of {@link Vocabulary}.* */
  List<Vocabulary> findVocabularies();

  Long participationCount(DbCohort dbCohort);
}

File: api/src/main/java/org/pmiops/workbench/cohortreview/CohortReviewService.kt
package org.pmiops.workbench.cohortreview

import org.pmiops.workbench.cohortreview.util.PageRequest
import org.pmiops.workbench.db.model.Cohort
import org.pmiops.workbench.db.model.CohortAnnotationDefinition
import org.pmiops.workbench.db.model.CohortReview
import org.pmiops.workbench.db.model.ParticipantCohortAnnotation
import org.pmiops.workbench.db.model.ParticipantCohortStatus
import org.pmiops.workbench.db.model.Workspace
import org.pmiops.workbench.model.ModifyParticipantCohortAnnotationRequest
import org.pmiops.workbench.model.WorkspaceAccessLevel

interface CohortReviewService {

    /**
     * Find the [Cohort] for the specified cohortId.
     *
     * @param cohortId
     * @return
     */
    fun findCohort(cohortId: Long): Cohort

    /**
     * Validate that a workspace exists for the specified workspaceId, and set the workspace's. CDR
     * version in [org.pmiops.workbench.cdr.CdrVersionContext].
     *
     * @param workspaceNamespace
     * @param workspaceName
     * @param workspaceId
     */
    fun validateMatchingWorkspaceAndSetCdrVersion(
            workspaceNamespace: String,
            workspaceName: String,
            workspaceId: Long,
            requiredAccess: WorkspaceAccessLevel): Workspace

    fun enforceWorkspaceAccessLevel(
            workspaceNamespace: String, workspaceId: String, requiredAccess: WorkspaceAccessLevel): WorkspaceAccessLevel

    /**
     * Find the [CohortReview] for the specified cohortId and cdrVersionId.
     *
     * @param cohortId
     * @param cdrVersionId
     * @return
     */
    fun findCohortReview(cohortId: Long?, cdrVersionId: Long?): CohortReview

    /**
     * Find the [CohortReview] for the specified cohortReviewId.
     *
     * @param cohortReviewId
     * @return
     */
    fun findCohortReview(cohortReviewId: Long?): CohortReview

    /**
     * Find the [CohortReview] for the specified ns, firecloudName and cohortReviewId
     *
     * @param ns
     * @param firecloudName
     * @param cohortReviewId
     * @return
     */
    fun findCohortReview(ns: String, firecloudName: String, cohortReviewId: Long?): CohortReview

    /**
     * Delete the specified cohort review.
     *
     * @param cohortReview
     */
    fun deleteCohortReview(cohortReview: CohortReview)

    /**
     * Find the [CohortReview] for the specified ns and firecloudName.
     *
     * @param ns
     * @param firecloudName
     * @return
     */
    fun getRequiredWithCohortReviews(ns: String, firecloudName: String): List<CohortReview>

    /**
     * Save the specified [CohortReview].
     *
     * @param cohortReview
     * @return
     */
    fun saveCohortReview(cohortReview: CohortReview): CohortReview

    /**
     * Save the [CohortReview] as well as the collection of [ParticipantCohortStatus].
     *
     * @param cohortReview
     * @param participantCohortStatuses
     */
    fun saveFullCohortReview(
            cohortReview: CohortReview, participantCohortStatuses: List<ParticipantCohortStatus>)

    /**
     * Save the specified [ParticipantCohortStatus].
     *
     * @param participantCohortStatus
     * @return
     */
    fun saveParticipantCohortStatus(
            participantCohortStatus: ParticipantCohortStatus): ParticipantCohortStatus

    /**
     * Find the [ParticipantCohortStatus] for the specified cohortReviewId and participantId.
     *
     * @param cohortReviewId
     * @param participantId
     * @return
     */
    fun findParticipantCohortStatus(cohortReviewId: Long?, participantId: Long?): ParticipantCohortStatus

    /**
     * Find a list of [ParticipantCohortStatus] for the specified cohortReviewId, filtering and
     * paging.
     *
     * @param cohortReviewId
     * @param pageRequest
     * @return
     */
    fun findAll(cohortReviewId: Long?, pageRequest: PageRequest): List<ParticipantCohortStatus>

    /**
     * Find count for the specified cohortReviewId, filtering and paging.
     *
     * @param cohortReviewId
     * @param pageRequest
     * @return
     */
    fun findCount(cohortReviewId: Long?, pageRequest: PageRequest): Long?

    /**
     * Save the specified [ParticipantCohortAnnotation].
     *
     * @param cohortReviewId
     * @param participantCohortAnnotation
     * @return
     */
    fun saveParticipantCohortAnnotation(
            cohortReviewId: Long?, participantCohortAnnotation: ParticipantCohortAnnotation): ParticipantCohortAnnotation

    /**
     * Save the [ParticipantCohortAnnotation] for the specified participantId, annotationId and
     * [ModifyParticipantCohortAnnotationRequest].
     *
     * @param annotationId
     * @param participantId
     * @param modifyParticipantCohortAnnotationRequest
     * @return
     */
    fun updateParticipantCohortAnnotation(
            annotationId: Long?,
            cohortReviewId: Long?,
            participantId: Long?,
            modifyParticipantCohortAnnotationRequest: ModifyParticipantCohortAnnotationRequest): ParticipantCohortAnnotation

    /**
     * Find the [CohortAnnotationDefinition] for the specified cohortAnnotationDefinitionId.
     *
     * @param cohortAnnotationDefinitionId
     * @return
     */
    fun findCohortAnnotationDefinition(cohortAnnotationDefinitionId: Long?): CohortAnnotationDefinition

    /**
     * Delete the [ParticipantCohortAnnotation] for the specified annotationId, cohortReviewId
     * and participantId.
     *
     * @param annotationId
     * @param cohortReviewId
     * @param participantId
     */
    fun deleteParticipantCohortAnnotation(
            annotationId: Long?, cohortReviewId: Long?, participantId: Long?)

    /**
     * Find the [ParticipantCohortAnnotation] for the specified cohortReviewId,
     * cohortAnnotationDefinitonId and participantId.
     *
     * @param cohortReviewId
     * @param cohortAnnotationDefinitionId
     * @param participantId
     * @return
     */
    fun findParticipantCohortAnnotation(
            cohortReviewId: Long?, cohortAnnotationDefinitionId: Long?, participantId: Long?): ParticipantCohortAnnotation

    /**
     * Find a list of [ParticipantCohortAnnotation] for the specified cohortReviewId and
     * participantId.
     *
     * @param cohortReviewId
     * @param participantId
     * @return
     */
    fun findParticipantCohortAnnotations(
            cohortReviewId: Long?, participantId: Long?): List<ParticipantCohortAnnotation>
}
